File: platform/core/src/main/java/org/openmole/ide/core/PropertyPanel.java
/*
 * Copyright (C) 2011 Mathieu leclaire <mathieu.leclaire at openmole.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.ide.core;

import java.awt.Component;
import java.util.HashMap;
import java.util.Map;
import javax.swing.JList;
import javax.swing.JPanel;
import javax.swing.plaf.basic.BasicComboBoxRenderer;
import org.openmole.ide.core.commons.Constants;
import org.openmole.ide.core.palette.TaskDataProxyFactory;
import org.openmole.ide.core.palette.SamplingDataProxyFactory;
import org.openmole.ide.core.palette.PrototypeDataProxyFactory;
import org.openmole.ide.core.palette.EnvironmentDataProxyFactory;
import org.openmole.ide.core.palette.IDataProxyFactory;
import org.openmole.ide.core.palette.ElementFactories;
import org.openmole.ide.core.control.MoleScenesManager;
import org.openmole.ide.core.palette.DataProxyUI;
import org.openmole.ide.core.properties.IPanelUI;
import org.openmole.ide.core.properties.IDataUI;
import scala.collection.JavaConversions;
import scala.collection.JavaConversions.*;

public class PropertyPanel extends javax.swing.JPanel {

    private static PropertyPanel instance;
    private DataProxyUI<? extends IDataUI> currentDataProxyUI;
    private IPanelUI<? extends IDataUI> currentPanelUI;
    private Boolean initEntity = false;
    private Map<String, Boolean> locked = new HashMap<String, Boolean>();

    public PropertyPanel() {
        initComponents();
        locked.put(Constants.TASK(), false);
        locked.put(Constants.PROTOTYPE(), false);
        locked.put(Constants.SAMPLING(), false);
        locked.put(Constants.ENVIRONMENT(), false);

        typeComboBox.setRenderer(new ItemRenderer());
        entityPanelScrollPane.setVisible(false);
    }

    public String getNameTextField() {
        return nameTextField.getText();
    }

//    public void save() {
//        if (!initEntity) {
//            if (currentDataUI != null) {
//                if (currentPanelUI != null) {
//                    DataUI pdata = currentPanelUI.saveContent();
//                    pdata.name_$eq(nameTextField.getText());
//                    currentDataUI.entity().updateDataUI(pdata);
//                }
//            }
//        }
//    }
    public void displayCurrentEntity() {
        displayCurrentEntity(currentDataProxyUI);
    }

    public void displayCurrentEntity(DataProxyUI<? extends IDataUI> dpu) {
        currentDataProxyUI = dpu;
        currentPanelUI = dpu.dataUI().buildPanelUI();

        nameTextField.setText(dpu.dataUI().name());
        typeComboBox.removeAllItems();
       // currentPanelUI.loadContent(currentDataProxyUI.dataUI());
        updateViewport((JPanel) currentPanelUI);
        entityPanelScrollPane.setVisible(true);

        locked.put(dpu.dataUI().entityType(), false);
        initEntity = false;
    }

    public void updateViewport(JPanel panel) {
        entityPanelScrollPane.getViewport().removeAll();
        entityPanelScrollPane.setViewportView(panel);

    }

    private void save() {
//        if (initEntity) {
//            if (nameTextField.getText().length() != 0) {
//                IDataProxyFactory currentModelElementFactory = (IDataProxyFactory) (typeComboBox.getSelectedItem());
//            //    DataProxyUI<? extends IDataUI> element = currentModelElementFactory.buildDataProxyUI(nameTextField.getText());
//                MoleSceneTopComponent.getDefault().refreshPalette();
//                displayCurrentEntity(currentModelElementFactory.buildDataProxyUI(nameTextField.getText()));
//                locked.put(currentModelElementFactory.factory().entityType(), false);
//            }
//        } else {
//            if (currentDataProxyUI != null) {
//                if (currentPanelUI != null) {
//                    currentDataProxyUI.updataDataUI(currentPanelUI.saveContent(nameTextField.getText()));
//                    currentDataProxyUI.dataUI_$eq(currentPanelUI.saveContent(nameTextField.getText()));
//                }
//            }
//        }
//        MoleSceneTopComponent.getDefault().refreshPalette();
//        initEntity = false;
    }

    private void initNewEntity(String entityType) {

        if (entityPanelScrollPane.getViewport() != null) {
            entityPanelScrollPane.getViewport().removeAll();
            //   nameTextField.setText(MoleScenesManager.getName(entityType, locked.get(entityType)));
            nameTextField.setText(MoleScenesManager.incrementCounter(entityType));
        }
        initEntity = true;

    }

    private void cleanBox() {
        typeComboBox.setEnabled(true);
        typeComboBox.removeAllItems();
    }

    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        jButton1 = new javax.swing.JButton();
        nameLabel = new javax.swing.JLabel();
        nameTextField = new javax.swing.JTextField();
        typeLabel = new javax.swing.JLabel();
        typeComboBox = new javax.swing.JComboBox();
        entityPanelScrollPane = new javax.swing.JScrollPane();
        saveButton = new javax.swing.JButton();
        jPanel1 = new javax.swing.JPanel();
        environmentToggleButton = new javax.swing.JToggleButton();
        prototypeToggleButton = new javax.swing.JToggleButton();
        taskToggleButton = new javax.swing.JToggleButton();
        samplingToggleButton = new javax.swing.JToggleButton();
        cancelButton = new javax.swing.JButton();

        jButton1.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.jButton1.text")); // NOI18N

        nameLabel.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.nameLabel.text")); // NOI18N

        nameTextField.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.nameTextField.text")); // NOI18N

        typeLabel.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.typeLabel.text")); // NOI18N

        typeComboBox.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                typeComboBoxActionPerformed(evt);
            }
        });

        saveButton.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.saveButton.text")); // NOI18N
        saveButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                saveButtonActionPerformed(evt);
            }
        });

        environmentToggleButton.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.environmentToggleButton.text")); // NOI18N
        environmentToggleButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                environmentToggleButtonActionPerformed(evt);
            }
        });

        prototypeToggleButton.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.prototypeToggleButton.text")); // NOI18N
        prototypeToggleButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                prototypeToggleButtonActionPerformed(evt);
            }
        });

        taskToggleButton.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.taskToggleButton.text")); // NOI18N
        taskToggleButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                taskToggleButtonActionPerformed(evt);
            }
        });

        samplingToggleButton.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.samplingToggleButton.text")); // NOI18N
        samplingToggleButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                samplingToggleButtonActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout jPanel1Layout = new javax.swing.GroupLayout(jPanel1);
        jPanel1.setLayout(jPanel1Layout);
        jPanel1Layout.setHorizontalGroup(
            jPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(jPanel1Layout.createSequentialGroup()
                .addComponent(taskToggleButton)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(prototypeToggleButton)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(samplingToggleButton)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(environmentToggleButton, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                .addContainerGap())
        );

        jPanel1Layout.linkSize(javax.swing.SwingConstants.HORIZONTAL, new java.awt.Component[] {environmentToggleButton, prototypeToggleButton, samplingToggleButton, taskToggleButton});

        jPanel1Layout.setVerticalGroup(
            jPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(jPanel1Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(jPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.CENTER)
                    .addComponent(taskToggleButton)
                    .addComponent(prototypeToggleButton)
                    .addComponent(samplingToggleButton)
                    .addComponent(environmentToggleButton))
                .addContainerGap())
        );

        cancelButton.setText(org.openide.util.NbBundle.getMessage(PropertyPanel.class, "PropertyPanel.cancelButton.text")); // NOI18N
        cancelButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                cancelButtonActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(this);
        this.setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addGroup(layout.createSequentialGroup()
                        .addComponent(jPanel1, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                        .addContainerGap())
                    .addGroup(layout.createSequentialGroup()
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                            .addComponent(entityPanelScrollPane, javax.swing.GroupLayout.Alignment.LEADING, javax.swing.GroupLayout.DEFAULT_SIZE, 421, Short.MAX_VALUE)
                            .addGroup(layout.createSequentialGroup()
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                    .addComponent(nameLabel)
                                    .addComponent(typeLabel))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING, false)
                                    .addComponent(typeComboBox, 0, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                                    .addComponent(nameTextField, javax.swing.GroupLayout.PREFERRED_SIZE, 289, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING, false)
                                    .addComponent(cancelButton, 0, 0, Short.MAX_VALUE)
                                    .addComponent(saveButton, javax.swing.GroupLayout.PREFERRED_SIZE, 70, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addGap(3, 3, 3)))
                        .addGap(27, 27, 27))))
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(jPanel1, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(14, 14, 14)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.CENTER)
                    .addComponent(saveButton)
                    .addComponent(nameTextField, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.CENTER)
                    .addComponent(typeComboBox, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(cancelButton))
                .addGap(20, 20, 20)
                .addComponent(entityPanelScrollPane, javax.swing.GroupLayout.DEFAULT_SIZE, 254, Short.MAX_VALUE)
                .addContainerGap())
            .addGroup(javax.swing.GroupLayout.Alignment.CENTER, layout.createSequentialGroup()
                .addGap(122, 122, 122)
                .addComponent(typeLabel)
                .addGap(292, 292, 292))
            .addGroup(layout.createSequentialGroup()
                .addGap(86, 86, 86)
                .addComponent(nameLabel)
                .addGap(328, 328, 328))
        );
    }// </editor-fold>//GEN-END:initComponents

    private void prototypeToggleButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_prototypeToggleButtonActionPerformed
        // setButtons(true);
        cleanBox();
        for (PrototypeDataProxyFactory dpu : JavaConversions.asJavaIterable(ElementFactories.modelPrototypes())) {
            typeComboBox.addItem(dpu);
        }
        initNewEntity(Constants.PROTOTYPE());
    }//GEN-LAST:event_prototypeToggleButtonActionPerformed

    private void samplingToggleButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_samplingToggleButtonActionPerformed
        // setButtons(false);
        cleanBox();
        for (SamplingDataProxyFactory dpu : JavaConversions.asJavaIterable(ElementFactories.modelSamplings())) {
            typeComboBox.addItem(dpu);
        }
        initNewEntity(Constants.SAMPLING());
    }//GEN-LAST:event_samplingToggleButtonActionPerformed

    private void environmentToggleButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_environmentToggleButtonActionPerformed
        // setButtons(true);
        cleanBox();
        for (EnvironmentDataProxyFactory dpu : JavaConversions.asJavaIterable(ElementFactories.modelEnvironments())) {
            typeComboBox.addItem(dpu);
        }
        initNewEntity(Constants.ENVIRONMENT());
    }//GEN-LAST:event_environmentToggleButtonActionPerformed

    private void typeComboBoxActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_typeComboBoxActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_typeComboBoxActionPerformed

    private void taskToggleButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_taskToggleButtonActionPerformed
        // setButtons(false);for (IDataProxyFactory dpu : JavaConversions.asJavaIterable(ElementFactories.modelEnvironments)) {
        cleanBox();
        for (TaskDataProxyFactory dpu : JavaConversions.asJavaIterable(ElementFactories.modelTasks())) {
            typeComboBox.addItem(dpu);
        }
        initNewEntity(Constants.TASK());

    }//GEN-LAST:event_taskToggleButtonActionPerformed

    private void saveButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_saveButtonActionPerformed
        // TODO add your handling code here:
        save();
    }//GEN-LAST:event_saveButtonActionPerformed

    private void cancelButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_cancelButtonActionPerformed
        displayCurrentEntity();
    }//GEN-LAST:event_cancelButtonActionPerformed
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JButton cancelButton;
    private javax.swing.JScrollPane entityPanelScrollPane;
    private javax.swing.JToggleButton environmentToggleButton;
    private javax.swing.JButton jButton1;
    private javax.swing.JPanel jPanel1;
    private javax.swing.JLabel nameLabel;
    private javax.swing.JTextField nameTextField;
    private javax.swing.JToggleButton prototypeToggleButton;
    private javax.swing.JToggleButton samplingToggleButton;
    private javax.swing.JButton saveButton;
    private javax.swing.JToggleButton taskToggleButton;
    private javax.swing.JComboBox typeComboBox;
    private javax.swing.JLabel typeLabel;
    // End of variables declaration//GEN-END:variables

    class ItemRenderer extends BasicComboBoxRenderer {
        @Override
        public Component getListCellRendererComponent(JList list, Object value, int index, boolean isSelected, boolean cellHasFocus) {
            super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus);
            if (value != null) {
                    setText(((IDataProxyFactory) value).factory().displayName());
            }
            return this;
        }
    }

    public static synchronized PropertyPanel getDefault() {
        if (instance == null) {
            instance = new PropertyPanel();
        }
        return instance;
    }
}

File: platform/core/org.openmole.ide.core.implementation/src/main/scala/org/openmole/ide/core/implementation/display/PropertyPanel.scala
/*
 * Copyright (C) 2011 <mathieu.leclaire at openmole.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.ide.core.implementation.display

import scala.swing._
import swing.Swing._
import swing.ListView._
import javax.swing.ImageIcon
import javax.swing.JPanel
import org.openmole.ide.core.model.panel.IPanelUI
//import org.openmole.ide.misc.image._
import org.openmole.ide.misc.widget.MigPanel
import scala.swing._
import org.openmole.ide.core.model.commons.IOType
import org.openmole.ide.core.model.dataproxy.IDataProxyUI
import org.openmole.ide.core.implementation.action._
import org.openmole.ide.core.implementation.display._
import scala.swing.event.ButtonClicked
import org.openmole.ide.core.model.dataproxy.ITaskDataProxyUI
import org.openmole.ide.misc.widget.MenuToggleButton
import org.openmole.ide.misc.widget.PopupMenu
import org.openmole.ide.core.model.commons.Constants._

class PropertyPanel extends MigPanel("fillx,wrap 4","[][grow,fill][][]", "[fill]30[]rel[grow,fill]rel[grow,fill]"){
//  Displays.currentType = TASK
//  var editable = true
//    
////  val environmentMenu = new Menu("Environment")
////  EnvironmentDisplay.implementationClasses.toList.sortBy(_.factory.displayName).foreach(
////    d => environmentMenu.contents += new MenuItem(new EnvironmentDisplayAction(d, ENVIRONMENT))
////  )
//  
////  val taskMenu = new Menu("Task")
////  TaskDisplay.implementationClasses.toList.sortBy(_.factory.displayName).foreach(
////    d => taskMenu.contents += new MenuItem(new TaskDisplayAction(d, TASK))
////  )
////  val prototypeMenu = new Menu("Prototype")
////  PrototypeDisplay.implementationClasses.toList.sortBy(_.factory.displayName).foreach(
////    d => prototypeMenu.contents += new MenuItem(new PrototypeDisplayAction(d, PROTOTYPE))
////  )
////  
////  val samplingMenu = new Menu("Sampling")
////  SamplingDisplay.implementationClasses.toList.sortBy(_.factory.displayName).foreach(
////    d => samplingMenu.contents += new MenuItem(new SamplingDisplayAction(d, SAMPLING))
////  )
//    
//  val fakeToggleButton = new MenuToggleButton(Some(new ImageIcon(ImageTool.loadImage("img/empty.png",30,30))))
//  val saveButton = new Button("Apply")
//  val cancelButton = new Button("Cancel")
//  val nameTextField = new TextField(15) 
////  val menuBar = new MenuBar{contents.append(prototypeMenu,taskMenu,samplingMenu,environmentMenu)}
// // menuBar.minimumSize = new Dimension(size.width,30)
//  
//  listenTo(saveButton,cancelButton)
//  reactions += {
//    case ButtonClicked(`saveButton`) =>  save
//    case ButtonClicked(`cancelButton`) =>  cancel}
//
//  val propertyScrollPane = new ScrollPane{verticalScrollBarPolicy = ScrollPane.BarPolicy.AsNeeded}
//  var taskPropertyComponent = new ScrollPane{verticalScrollBarPolicy = ScrollPane.BarPolicy.Never}
// // contents+= (menuBar,"span 4, growx")
//  contents+= fakeToggleButton
//  contents+= nameTextField
//  contents+= saveButton
//  contents+= cancelButton
//  
//  contents+= (propertyScrollPane,"span 4,growx, growy")
//  contents+= (taskPropertyComponent,"span 4,growx")
//  
//  def displayCurrentTypeIcon: Unit = displayCurrentTypeIcon(Displays.dataProxy.get)
//  
//  def displayCurrentTypeIcon(p: IDataProxyUI):Unit = {
////    contents.remove(1) 
////    contents.insert(1,new MenuToggleButton(
////        Some(new ImageIcon(ImageTool.loadImage(p.dataUI.imagePath,30,30))))
////      {popup= Displays.firstManagementMenu})
////    revalidate
//  }
//  
//  def displayCurrentEntity: Unit = displayCurrentEntity(Displays.dataProxy.get)
//  
//  def displayCurrentEntity(p: IDataProxyUI): Unit = {
////    nameTextField.text = p.dataUI.name
////    val pui = Displays.buildPanelUI
////    updateViewport(pui.peer)
////    displayCurrentTypeIcon(p)
////    
////    pui.peer.getComponents.foreach(_.setEnabled(editable))
////    pui.peer.getComponents.foreach(_.setEnabled(editable))
////    peer.getComponents.foreach(_.setEnabled(editable))
////    menuBar.peer.getComponents.foreach(_.setVisible(editable))
//  }
//  
//  def cleanViewport = {
////    removeViewport
////    nameTextField.text = ""
////    repaint
////    revalidate
//  }
//  
//  def removeViewport = {
////    propertyScrollPane.peer.getViewport.removeAll
////    contents.remove(1) 
////    contents.insert(1,fakeToggleButton)
//    }
//    
//  
//  def updateViewport(panel: JPanel)= {
////    removeViewport
////    updateTaskViewport
////    propertyScrollPane.peer.setViewportView(panel)
////    revalidate
//  }
//  
//  def updateTaskViewport = {
////    taskPropertyComponent.peer.getViewport.removeAll
////    Displays.dataProxy match {
////      case Some(x: ITaskDataProxyUI)=> {
////          taskPropertyComponent.peer.setViewportView(buildTaskPropertyPanel(x).peer)
////          taskPropertyComponent.visible =  true
////        }
////      case _=> taskPropertyComponent.visible = false
////    }
////    taskPropertyComponent.revalidate
//  }
//  
//  def save = {
//    Displays.saveContent
//    Displays.initMode = false
//  }
//  
//  def cancel = displayCurrentEntity
//  
//  def initNewEntity = {
//    Displays.initMode = true
//    nameTextField.text = ""
//   // updateViewport(Displays.buildPanelUI.peer)
//    displayCurrentTypeIcon
//  }
//  
// 
//  def buildTaskPropertyPanel(dpu: ITaskDataProxyUI) = {
////    var empty = true
////    val mp = new MigPanel("wrap 4","[]5[]","[]5[]") {
////      if (dpu.dataUI.prototypesIn.size>0){
////        empty= false
////        contents+= new Label("In")
////        contents+= (new MigPanel(""){
////            dpu.dataUI.prototypesIn.foreach(p=> contents+= 
////                                            secondManagementMenu(p.dataUI.imagePath,p.dataUI.name,PrototypeDisplay.secondManagementMenu(dpu,p,IOType.INPUT)))},"wrap")}
////      if (dpu.dataUI.prototypesOut.size>0){
////        empty= false
////        contents+= new Label("Out")
////        contents+= (new MigPanel(""){
////            dpu.dataUI.prototypesOut.foreach(p=> contents+= 
////                                             secondManagementMenu(p.dataUI.imagePath,p.dataUI.name,PrototypeDisplay.secondManagementMenu(dpu,p,IOType.OUTPUT)))},"wrap")}
////      if (dpu.dataUI.sampling.isDefined) {
////        empty= false
////        contents+= new Label("Sampling")
////        contents+= new MigPanel(""){contents+= secondManagementMenu(dpu.dataUI.sampling.get.dataUI.imagePath,
////                                                                    dpu.dataUI.sampling.get.dataUI.name,
////                                                                    SamplingDisplay.secondManagementMenu(dpu,dpu.dataUI.sampling.get))}}
//////      if (dpu.dataUI.environment.isDefined) {
//////        empty= false
//////        contents+= new Label("Environment")
//////        contents+= new MigPanel(""){contents+= secondManagementMenu(dpu.dataUI.environment.get.dataUI.imagePath,
//////                                                                    dpu.dataUI.environment.get.dataUI.name,
//////                                                                    EnvironmentDisplay.secondManagementMenu(dpu,dpu.dataUI.environment.get))}}
////    }
////    taskPropertyComponent.visible = !empty
////    if (empty) taskPropertyComponent.size.height = 0
////    mp.peer.getComponents.foreach(_.setEnabled(editable))
////    mp
//  }
//     
////  def secondManagementMenu(impath: String, displayname: String,p: PopupMenu) = new MenuToggleButton(
////    Some(new ImageIcon(ImageTool.loadImage(impath,30,30))),displayname){popup= p; popup.peer.getComponents.foreach(_.setEnabled(editable))}
}
